src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

```diff
@@ -252,19 +254,135 @@ internal static IEnumerable<IEdmOperation> FilterBoundOperationsWithSameTypeHier
-            IEnumerable<IEdmOperation> exactMatches = functions.Where(f => f.Parameters.Count() == parameters.Count() + (f.IsBound ? 1 : 0));
+            IEnumerable<IEdmOperation> exactMatches = functions.Where(f => f.Parameters.Count() == parameterCount + (f.IsBound ? 1 : 0));
             return exactMatches.Count() > 0 ? exactMatches : functions;
         }
 
+        /// <summary>
+        /// Filters the operations by parameter count.
+        /// </summary>
+        /// <param name="functions">The operations.</param>
+        /// <param name="parameterCount">The count of non-binding parameter names to match.</param>
+        /// <returns>The best matching operations based on parameters.</returns>
+        internal static IList<IEdmOperation> FindBestOverloadBasedOnParameterCount(this IList<IEdmOperation> functions, int parameterCount)
+        {
+            IList<IEdmOperation> exactMatches = null;
+
+            // The best match out of a list of candidates is the one that has the same number of (non-binding) parameters as specified.
+            for (int i = 0; i < functions.Count; i++)
+            {
+                if (functions[i].Parameters.Count() == parameterCount + (functions[i].IsBound ? 1 : 0))
+                {
+                    if (exactMatches == null)
+                    {
+                        exactMatches = new List<IEdmOperation>();
+                    }
+
+                    exactMatches.Add(functions[i]);
+                }
+            }
+
+            return exactMatches == null ? functions : exactMatches;
+        }
+
+        /// <summary>
+        /// Filters the operations candidates based on parameters.
+        /// </summary>
+        /// <param name="operations">The operations.</param>
+        /// <param name="bindingType">Binding type for the operation.</param>
+        /// <param name="parameterNameList">The parameters.</param>
+        /// <param name="caseInsensitive">Whether case insensitive.</param>
+        /// <param name="actionItems">outputs the actions found while matching functions</param>
+        /// <returns>Operations filtered by parameter.</returns>
+        internal static IList<IEdmOperation> FilterOperationCandidatesBasedOnParameterList(this IEnumerable<IEdmOperation> operations, IEdmType bindingType, IList<string> parameterNameList, bool caseInsensitive, out IList<IEdmOperation> actionItems)
+        {
+            Debug.Assert(operations != null, "operations");
+            Debug.Assert(parameterNameList != null, "parameters");
+
+            bool hasParameters = parameterNameList.Count > 0;
+            actionItems = null;
+            IList<IEdmOperation> operationCandidates = new List<IEdmOperation>();
+
+            foreach (var operation in operations)
+            {
+                if (bindingType != null)
+                {
+                    operation.EnsureOperationBoundWithBindingParameter();

```

Comment:
>EnsureOperationBoundWithBindingParameter [](start = 30, length = 40)

should this also validate that the binding type is the same as the first parameter, or is that done elsewhere?

Reply:
Yes, although I don't like it as much but the calling function calls this extension method only with already filtered candidates if a binding type exists. 

---
In reply to: [370730518](https://github.com/OData/odata.net/pull/1602#discussion_r370730518) [](ancestors = 370730518)